internal/evmreader/evmreader.go

```diff
@@ -0,0 +1,344 @@
+	"context"
+	"fmt"
+	"log/slog"
+	"math/big"
+
+	"github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/cartesi/rollups-node/pkg/contracts/inputbox"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type (
+	Address            = common.Address
+	Application        = model.Application
+	Input              = model.Input
+	DefaultBlock       = model.DefaultBlock
+	InputBoxInputAdded = inputbox.InputBoxInputAdded
+	FilterOpts         = bind.FilterOpts
+	Context            = context.Context
+	Header             = types.Header
+	Subscription       = ethereum.Subscription
+)
+
+// Interface for Input reading
+type InputSource interface {
+	// Wrapper for FilterInputAdded(), which is automatically generated by go-ethereum and cannot be used for testing
+	RetrieveInputs(
+		opts *FilterOpts,
+		appContract []Address,
+		index []*big.Int,
+	) ([]InputBoxInputAdded, error)
+}
+
+// Interface for the node repository
+type EvmReaderRepository interface {
+	InsertInputsAndUpdateLastProcessedBlock(
+		ctx Context,
+		inputs []Input,
+		blockNumber uint64,
+		appAddress Address,
+	) error
+	GetAllRunningApplications(
+		ctx context.Context,
+	) ([]Application, error)
+}
+
+// EthClient mimics part of ethclient.Client functions to narrow down the
+// interface needed by the EvmReader. It must be bound to an HTTP endpoint
+type EthClient interface {
+	HeaderByNumber(
+		ctx Context,
+		number *big.Int,
+	) (*Header, error)
+}
+
+// EthWsClient mimics part of ethclient.Client functions to narrow down the
+// interface needed by the EvmReader. It must be bound to a WS endpoint
+type EthWsClient interface {
+	SubscribeNewHead(
+		ctx Context,
+		ch chan<- *Header,
+	) (Subscription, error)
+}
+
+type SubscriptionError struct {
+	Cause error
+}
+
+func (e *SubscriptionError) Error() string {
+	return fmt.Sprintf("Subscription error : %v", e.Cause)
+}
+
+type nodeConfig struct {
+	inputBoxBlockNumber uint64
+	defaultBlock        DefaultBlock
+}
+
+// EvmReader reads inputs from the blockchain
+type EvmReader struct {
+	client      EthClient
+	wsClient    EthWsClient
+	inputSource InputSource
+	repository  EvmReaderRepository
+	config      nodeConfig
+}
+
+func (r *EvmReader) String() string {
+	return "evm-reader"
+}
+
+// Creates a new EvmReader
+func NewEvmReader(
+	client EthClient,
+	wsClient EthWsClient,
+	inputSource InputSource,
+	repository EvmReaderRepository,
+	inputboxBlockNumber uint64,
+	defaultBlock DefaultBlock,
+
+) EvmReader {
+	return EvmReader{
+		client:      client,
+		wsClient:    wsClient,
+		inputSource: inputSource,
+		repository:  repository,
+		config: nodeConfig{
+			inputBoxBlockNumber: inputboxBlockNumber,
+			defaultBlock:        defaultBlock,
+		},
+	}
+}
+
+func (r *EvmReader) Start(
+	ctx context.Context,
+	ready chan<- struct{},
+) error {
```

Comment:
Both the `EvmReader` and `EvmReaderService` utilize the `Start` services interface. Is there any case we call `EvmReader` directly? I find it odd reading through the code that you start the service, then the service starts the evm reader. It implies the evm reader can exist without the service, to be started directly by someone else.

Reply:
In fact, `EvmReader` is not a service. So `Start()` should be dropped.